### PR TITLE
Fix for pip install of dist tar

### PR DESCRIPTION
**Pull request by davidolrik**
Running `pip install --no-deps file:///home/username/pysodium-0.7.5.linux-x86_64.tar.gz` fails, because the `README.md` is not included in the tar.

This is also the method that Poetry uses, hence you can’t add pysodium to a project using Poetry.

This small fix simply ignores `FileNotFoundError`.

### CODE REVIEW

**Comment by stef** [comment]
i don't quite get this. what is the source of the tarfile?
and why does the opening of the file with a postfixed "FIXME" string improve things?

**Comment by davidolrik** [comment]
Sorry - The `FIXME` shouldn't be there, it was a leftover from my local testing and has now been removed.

The source of the tar is [pypi.org](https://pypi.org/project/pysodium/#files), and in order for me to test the exception from poetry I needed to force the error.

Before this fix, when doing `poetry add pysodium` I would get an exception because `pip` could not find the readme in the tar it fetched from pypi.org.

**Comment by stef** [comment]
> the README.md is not included in the tar.
i just tried this: `curl -L https://files.pythonhosted.org/packages/2b/0f/d3cf795f0ac72740d8858b9620ea2d952525a2e7dafc8d4eb96269b2b0b6/pysodium-0.7.5.tar.gz | tar tzv`

and this is the abbreviated result:
```
drwxr-xr-x s/s               0 2020-01-31 15:33 pysodium-0.7.5/
-rw-r--r-- s/s             229 2017-07-22 10:27 pysodium-0.7.5/AUTHORS
-rw-r--r-- s/s            8076 2020-01-19 23:25 pysodium-0.7.5/README.md
...
```

it seems the file is there.

what exactly is this `poetry` thing?

**Comment by stef** [comment]
ok, i found poetry. i assume it is this? https://python-poetry.org/

**Comment by stef** [comment]
i also tried what you wrote as the initial problem:

```
 pip install --no-deps https://files.pythonhosted.org/packages/2b/0f/d3cf795f0ac72740d8858b9620ea2d952525a2e7dafc8d4eb96269b2b0b6/pysodium-0.7.5.tar.gz
Collecting https://files.pythonhosted.org/packages/2b/0f/d3cf795f0ac72740d8858b9620ea2d952525a2e7dafc8d4eb96269b2b0b6/pysodium-0.7.5.tar.gz
  Downloading pysodium-0.7.5.tar.gz (19 kB)
Building wheels for collected packages: pysodium
  Building wheel for pysodium (setup.py) ... done
  Created wheel for pysodium: filename=pysodium-0.7.5-py3-none-any.whl size=19220 sha256=472a5686498c15102fbb4849b1d20f341bd12ad72f2fd67568865ba60ff45a97
  Stored in directory: /home/s/.cache/pip/wheels/b4/d4/f1/6737d976cc372bf72600e20f93d49fa55f85f65f2aec488803
Successfully built pysodium
Installing collected packages: pysodium
Successfully installed pysodium-0.7.5
```
there is something else wrong either with your install or poetry

**Comment by davidolrik** [comment]
It also works for me using the source tar file, and I'm currently using that as a workaround.

The tar file that is used is the platform specific one: `pysodium-0.7.5.linux-x86_64.tar.gz` which does not include anything other than the code.

**Comment by stef** [comment]
ok, so i did: 

```
pip install --no-deps https://files.pythonhosted.org/packages/33/e0/33879cc17ee1825be69f038bdbceedfe020255094eb62b53fcd01cbc2b26/pysodium-0.7.5.linux-x86_64.tar.gz
```

and indeed it complains, but not about `README.md` but about `setup.py` missing:

```
    ERROR: Command errored out with exit status 1:
     command: /tmp/env/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-42ficn0k/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-42ficn0k/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-qd7yy8nz
         cwd: /tmp/pip-req-build-42ficn0k/
    Complete output (5 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/env/lib/python3.8/tokenize.py", line 392, in open
        buffer = _builtin_open(filename, 'rb')
    FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pip-req-build-42ficn0k/setup.py'
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

but i suspect that file is not meant to be directly installed via `pip`, as when i look inside with
```sh
curl -L https://files.pythonhosted.org/packages/33/e0/33879cc17ee1825be69f038bdbceedfe020255094eb62b53fcd01cbc2b26/pysodium-0.7.5.linux-x86_64.tar.gz | tar tzv
```

there is indeed only the minimal included:

```
drwxr-xr-x s/s               0 2020-01-31 15:34 ./usr/
drwxr-xr-x s/s               0 2020-01-31 15:34 ./usr/lib/
drwxr-xr-x s/s               0 2020-01-31 15:34 ./usr/lib/python2.7/
drwxr-xr-x s/s               0 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/
drwxr-xr-x s/s               0 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/pysodium-0.7.5-py2.7.egg-info/
-rw-r--r-- s/s               1 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/pysodium-0.7.5-py2.7.egg-info/dependency_links.txt
-rw-r--r-- s/s             238 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/pysodium-0.7.5-py2.7.egg-info/SOURCES.txt
-rw-r--r-- s/s           10260 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/pysodium-0.7.5-py2.7.egg-info/PKG-INFO
-rw-r--r-- s/s              14 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/pysodium-0.7.5-py2.7.egg-info/top_level.txt
drwxr-xr-x s/s               0 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/pysodium/
-rw-r--r-- s/s           66212 2020-01-31 15:07 ./usr/lib/python2.7/site-packages/pysodium/__init__.py
-rw-r--r-- s/s           53160 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/pysodium/__init__.pyc
drwxr-xr-x s/s               0 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/test/
-rw-r--r-- s/s               1 2017-07-11 18:59 ./usr/lib/python2.7/site-packages/test/__init__.py
-rw-r--r-- s/s             136 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/test/__init__.pyc
-rw-r--r-- s/s           29472 2020-01-31 15:34 ./usr/lib/python2.7/site-packages/test/test_pysodium.pyc
100 36577  100 36577    0     0   435k      0 --:--:-- --:--:-- --:--:--  435k
-rw-r--r-- s/s           32314 2020-01-31 15:04 ./usr/lib/python2.7/site-packages/test/test_pysodium.py
```

why don't you simply do `pip install pysodium`?  why do you want to pip install something, that is apparently not meant to be installed by `pip` directly?

**Comment by davidolrik** [comment]
My fix is wrong - the issue is two fold, and lies in both poetry and in pysodium.

The file `pysodium-0.7.5.linux-x86_64.tar.gz` is broken and shouldn't have been uploaded to pypi.
When you do a `pip install pysodium`, it downloads the sdist, created a wheel and installs it.

When poetry tries to install pysodium it gets the `pysodium-0.7.5.linux-x86_64.tar.gz` by mistake and then fails.
I suggest that you delete `pysodium-0.7.5.linux-x86_64.tar.gz` from pypi.

**Comment by stef** [comment]
why is `pysodium-0.7.5.linux-x86_64.tar.gz` wrong? it's a bdist. those should not contain readmes and such. they contain only what's needed in a `pythonx.x/site-packages` directory, no? 

on another note, i thought that pypi is immutable, cannot delete stuff there.

**Comment by sinoroc** [comment]
> why is `pysodium-0.7.5.linux-x86_64.tar.gz` wrong? it's a bdist. those should not contain readmes and such. they contain only what's needed in a `pythonx.x/site-packages` directory, no?

Looks like `pysodium-0.7.5.linux-x86_64.tar.gz` was created with `python setup.py bdist`. I am not sure for what cases a _bdist_ is useful. But I am quite sure it is quite close to pointless to distribute those on _PyPI_. Since for example they contain absolute paths. I have not looked at the content much further.

On the other hand, it is very useful to upload the _sdist_ and the _wheel_. For example `python setup.py bdist_wheel upload`.

Not sure, if it is already known yet. Just in case, I would like to recommend reading those:

* [_Packaging your project_](https://packaging.python.org/guides/distributing-packages-using-setuptools/#packaging-your-project)
* [_Uploading your Project to PyPI_](https://packaging.python.org/guides/distributing-packages-using-setuptools/#uploading-your-project-to-pypi)

In short: use _twine_ to upload the _sdist_ and the _wheel_ of your project.

> on another note, i thought that pypi is immutable, cannot delete stuff there.

If I remember right you can delete, but you cannot replace with a different file of the same name. Nowadays the preferred operation in such cases is called [_yanking_](https://pypi.org/help/#yanked).

**Comment by bufke** [comment]
For any folks finding this via google, one work-around is to use Poetry 1.0.10. This problem only exists in Poetry 1.1.x.

**Comment by davidolrik** [comment]
If you want to stay on the latest poetry version, you can add pysodium directly like this:

```sh
poetry add git+https://github.com/stef/pysodium.git#0.7.5
```
